File: src/net/dinkla/raytracer/math/Vector3D.java
package net.dinkla.raytracer.math;

import static java.lang.Math.sqrt;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 10.04.2010
 * Time: 15:17:35
 * To change this template use File | Settings | File Templates.
 */
public class Vector3D extends Element3D {

    public static final Vector3D ZERO = new Vector3D(0, 0, 0);
    static public final Vector3D UP = new Vector3D(0, 1, 0);
    static public final Vector3D DOWN = new Vector3D(0, -1, 0);
    public static final Vector3D JITTER = new Vector3D(0.0072f, 1.0f, 0.0034f);
    public static final Vector3D BACK = new Vector3D(0, 0, -1);

    public Vector3D(final float x, final float y, final float z) {
        super(x, y, z);
    }

    public Vector3D(final Element3D e) {
        super(e);
    }

    public Vector3D plus(final Vector3D v) {
        return new Vector3D(x + v.x, y + v.y, z + v.z);
    }

    public Vector3D minus(final Vector3D v) {
        return new Vector3D(x - v.x, y - v.y, z - v.z);
    }

    public Vector3D mult(final float s) {
        return new Vector3D(s * x, s * y, s * z);
    }

    public float dot(final Vector3D v)  {
        return x * v.x + y * v.y + z * v.z;
    }

    public float dot(final Normal v)  {
        return x * v.x + y * v.y + z * v.z;
    }

    public Vector3D cross(Vector3D v)  {
        return new Vector3D(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
    }

    public Vector3D normalize() {
        final float l = length();
        return new Vector3D(x / l, y / l, z / l);
    }

    public Vector3D negate() {
        return new Vector3D(-x, -y, -z);
    }

}

File: src/main/kotlin/net/dinkla/raytracer/math/Vector3D.kt
package net.dinkla.raytracer.math

class Vector3D(x: Double, y: Double, z: Double) : Element3D(x, y, z) {

    constructor(x: Int, y: Int, z: Int) : this(x.toDouble(), y.toDouble(), z.toDouble())

    constructor(e: Element3D) : this(e.x, e.y, e.z)

    operator fun plus(v: Vector3D): Vector3D = Vector3D(x + v.x, y + v.y, z + v.z)

    operator fun minus(v: Vector3D): Vector3D = Vector3D(x - v.x, y - v.y, z - v.z)

    operator fun times(s: Double): Vector3D = Vector3D(s * x, s * y, s * z)

    infix fun dot(v: Vector3D): Double = x * v.x + y * v.y + z * v.z

    infix fun dot(v: Normal?): Double {
        return if (null == v) x * y * z
        else x * v.x + y * v.y + z * v.z
    }

    infix fun cross(v: Vector3D): Vector3D = Vector3D(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x)

    fun normalize(): Vector3D {
        val l = length()
        return Vector3D(x / l, y / l, z / l)
    }

    operator fun unaryMinus() = Vector3D(-x, -y, -z)

    fun volume(): Double = x * y * z

    companion object {
        val ZERO = Vector3D(0, 0, 0)
        val RIGHT = Vector3D(1, 0, 0)
        val UP = Vector3D(0, 1, 0)
        val DOWN = Vector3D(0, -1, 0)
        val FORWARD = Vector3D(0, 0, 1)
        val BACK = Vector3D(0, 0, -1)
        val JITTER = Vector3D(0.0072, 1.0, 0.0034)
    }

}

operator fun Double.times(v: Vector3D) = Vector3D(this * v.x, this * v.y, this * v.z)
